### PR TITLE
Nested horizontal virtualized list viewability

### PR DESCRIPTION
**Pull request by vincentvella**
## Summary

This Pull Request fixes issues with viewability events in the Virtualized List component when nesting in opposing directions.

Essentially, the VirtualizedList is currently only considering same-direction viewability. To fix that and stop the viewability computations from running on nested Flatlists, I needed a way to allow nested Flatlists to access the parent's currently visible elements. By storing the list's currently viewable items in-memory, I was able to reach into the parent's context and pass that data to the ViewabilityHelper of all of its children.

See #35180 

## Changelog

[General] [Fixed] - Nested horizontal virtualized list viewability events

## Test Plan

Working on adding some jest tests but there's some interesting stuff happening with react-test-renderer in the virtualized list related to viewability (likely need to trigger some onLayout events). Doesn't seem easy to test at this point.

https://user-images.githubusercontent.com/22749569/200435788-31abe231-3fc5-498d-979e-aa4f3f12798b.mp4

### CODE REVIEW

**Review by NickGerleman — CHANGES_REQUESTED** [review]
Thanks for the contribution!

The approach of keeping track of whether the outer cell is visible isn't quite correct I think, since the inner list can still be arbitrarily offset within the cell to not be visible itself. I have a PR internally where we nested lists are fully aware of their cross-axis position to account for this. It ended up adding complexity and I thought it may be worth pulling out the viewability logic to a unique component, but it seems like folks are repeatedly hitting this so it might be worth cleaning up again and submitting.

**Comment by vincentvella** [comment]
> since the inner list can still be arbitrarily offset within the cell to not be visible itself.

@NickGerleman is there any chance you have an example of what you're describing above? It'll be useful to help diagnose the issues with this PR. For my particular use-case this solves the issue I'm attempting to get past so I might use this as a temporary patch until your version/strategy is merged.

My thought process was if an inner cell was no longer visible or became not visible at a time post-render it'd fire one of the layout events and the list would recalculate its positioning but I'm interested to see if there's an edge case I'd missed

**Comment by vincentvella** [comment]
> since the inner list can still be arbitrarily offset within the cell to not be visible itself.

@NickGerleman is there any chance you have an example of what you're describing above? It'll be useful to help diagnose the issues with this PR. For my particular use-case this solves the issue I'm attempting to get past so I might use this as a temporary patch until your version/strategy is merged.

My thought process was if an inner cell was no longer visible or became not visible at a time post-render it'd fire one of the layout events and the list would recalculate its positioning but I'm interested to see if there's an edge case I'd missed

**Comment by NickGerleman** [comment]
In the below example, cell 2 is visible, but none of the children in the nested list are yet visible in the horizontal axis. So deriving visibility from whether the list is in a visible cell doesn't work quite generically.

A solution you can use to do that accurately is to instead thread both main-axis and cross-axis position down the child contexts. That allows individual lists to know their viewport coverage in both axis. It means requirements like "must be 20% visible" can also apply to the cross-axis as well.

Last week has been hectic, but I can export the change as a GitHub PR and retest it soon. Part of why I was holding off is that this change will need more UTs than what I had added, or the nested RNTester example you discovered. 

image

**Comment by vincentvella** [comment]
> 

This makes total sense, I'm happy to help with any unit testing if you are able to export the PR to Github!
I attempted to do something similar but I was attempting to reach up from the child to get the scroll positioning. I think it may have only been grabbing the outermost parent because it wasn't firing the onViewableItemsChanged appropriately anymore.

**Pull request closed without merging**
